src/logging.c

```diff
@@ -99,49 +149,97 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
     g_rcutils_logging_output_handler = &rcutils_logging_console_output_handler;
     g_rcutils_logging_default_logger_level = RCUTILS_DEFAULT_LOGGER_DEFAULT_LEVEL;
 
-    // Check the environment variable for line buffered output
-    const char * line_buffered;
+    const char * line_buffered = NULL;
     const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED", &line_buffered);
-
     if (NULL == ret_str) {
-      if (strcmp(line_buffered, "1") == 0) {
-        g_force_stdout_line_buffered = true;
-      } else if (strcmp(line_buffered, "0") != 0 && strcmp(line_buffered, "") != 0) {
+      if (strcmp(line_buffered, "") != 0) {
         fprintf(
           stderr,
-          "Warning: unexpected value [%s] specified for RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED. "
-          "Default value 0 will be used. Valid values are 1 or 0.\n",
-          line_buffered);
+          "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED is now ignored.  "
+          "Please set RCUTILS_LOGGING_USE_STDOUT and RCUTILS_LOGGING_BUFFERED_STREAM "
+          "to control the stream and the buffering of log messages.\n");
       }
     } else {
       fprintf(
-        stderr, "Error getting env. variable "
+        stderr, "Error getting environment variable "
         "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s\n", ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Error getting environment variable RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s", ret_str);
+      return RCUTILS_RET_ERROR;
+    }
+
+    // Set the default output stream for all severities to stderr so that errors
+    // are propagated immediately.
+    // The user can choose to set the output stream to stdout by setting the
+    // RCUTILS_LOGGING_USE_STDOUT environment variable to 1.
+    enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_LOGGING_USE_STDOUT", "use stderr",
+      "use stdout");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_EMPTY:
+      case RCUTILS_GET_ENV_ZERO:
+        g_output_stream = stderr;
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        g_output_stream = stdout;
+        break;
     }
 
-    // Check the environment variable for colorized output
-    const char * colorized_output;
-    ret_str = rcutils_get_env("RCUTILS_COLORIZED_OUTPUT", &colorized_output);
+    // Allow the user to choose how buffering on the stream works by setting
+    // RCUTILS_LOGGING_BUFFERED_STREAM.
+    // With an empty environment variable, use the default of the stream.
+    // With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be line buffered.
+    retval =
+      rcutils_get_env_var_zero_or_one(
+      "RCUTILS_LOGGING_BUFFERED_STREAM", "not buffered",
+      "buffered");
```

Comment:
nitpick: this is weird indentation I would expect:

```c
retval = rcutils_get_env_var_zero_or_one(
  ...);
```

or

```c
retval =
  rcutils_get_env_var_zero_or_one(...);
```

But not both...

If you must do it like this, I don't know why the linter doesn't make you do:

```c
retval =
  rcutils_get_env_var_zero_or_one(
    ...);
```

Reply:
Again, I have no idea why cpplint and uncrustify ask for the things they do.  I write code, and then I fight them until they pass.  I can fight them some more to make it seem more like real code.

Reply:
There is nothing weird about the above. An indentation happens for every opened parenthesis. The fact that `retval =` and `rcutils_get_env_var_zero_or_one(` are on two separate lines doesn't change the fact that the following arguments are within parenthesis and therefore one level indented. What William was pointing out was that having `rcutils_get_env_var_zero_or_one(` on the same level as the arguments isn't desired since it makes the code hard to visually parse. The first snipped William posted will pass the linter just fine and is what we use throughout the code.

Reply:
Honestly I don't care that much, I just thought I'd mention it in case it was unintentional.

I don't think what's there now should pass the linter, as for me things inside parenthesis should be indented once _from the line_ that the parenthesis open on, not the line the statement starts on. But again, I'm not going to waste time on it.